src/meta/src/hummock/compaction_schedule_policy.rs

```diff
@@ -0,0 +1,722 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::sync::Arc;
+
+use rand::Rng;
+use risingwave_hummock_sdk::HummockContextId;
+use risingwave_pb::hummock::{CompactTask, SubscribeCompactTasksResponse};
+use tokio::sync::mpsc::Receiver;
+
+use super::{Compactor, HummockManager};
+use crate::storage::MetaStore;
+use crate::MetaResult;
+
+const STREAM_BUFFER_SIZE: usize = 4;
+
+/// The implementation of compaction task scheduling policy.
+#[async_trait::async_trait]
+pub trait CompactionSchedulePolicy: Send + Sync {
+    /// Get next idle compactor to assign task.
+    async fn next_idle_compactor<S: MetaStore>(
+        &mut self,
+        hummock_manager: &HummockManager<S>,
+        compact_task: Option<&CompactTask>,
+    ) -> Option<Arc<Compactor>>;
+
+    /// Get next compactor to assign task.
+    fn next_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>>;
+
+    fn random_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>>;
+
+    fn add_compactor(
+        &mut self,
+        context_id: HummockContextId,
+        max_concurrent_task_number: u64,
+    ) -> Receiver<MetaResult<SubscribeCompactTasksResponse>>;
+
+    fn remove_compactor(&mut self, context_id: HummockContextId);
+
+    /// Notify the `CompactorManagerInner` of the completion of a task to adjust the next compactor
+    /// to schedule.
+    fn report_compact_task(&mut self, context_id: HummockContextId, task: &CompactTask);
+}
+
+pub enum CompactionSchedulePolicyImpl {
+    RoundRobin(RoundRobinPolicy),
```

Comment:
This operation is not frequently enough, so it is no necessary to use `enum` instead of `dynamic`

Reply:
I cannot use `Box<dyn ...>`, because `next_idle_compactor` is a generic method, making any implementation of `CompactionSchedulePolicy` not object safe.